File: core/cc/encoder.cpp
/*
 * Copyright (C) 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "encoder.h"
#include "schema.h"
#include "stream_writer.h"

#include <cstring>

namespace core {

Encoder::Encoder(std::shared_ptr<StreamWriter> output) : mOutput(output), mLastObjectId(0) {
    mEntities.insert(std::make_pair(nullptr, 0));
}

void Encoder::Bool(bool v) {
    uint8_t b = v ? 1 : 0;
    mOutput->write(&b, 1);
}

void Encoder::Int8(int8_t v) {
    mOutput->write(&v, 1);
}

void Encoder::Uint8(uint8_t v) {
    mOutput->write(&v, 1);
}

void Encoder::Uint16(uint16_t v) {
    uint8_t buf[9];
    uint16_t space = 0x7f;
    uint8_t tag = 0;
    for (int o = 8; true; o--) {
        if (v <= space) {
            buf[o] = uint8_t(v) | uint8_t(tag);
            mOutput->write(buf + o, 9 - o);
            return;
        }
        buf[o] = uint8_t(v);
        v = v >> 8;
        space >>= 1;
        tag = (tag >> 1) | 0x80;
    }
}

void Encoder::Int16(int16_t v) {
    uint16_t uv = uint16_t(v) << 1;
    Uint16((v < 0) ? ~uv : uv);
}

void Encoder::Float32(float v) {
    uint32_t bits = *reinterpret_cast<uint32_t*>(&v);
    uint32_t shuffled =
            ((bits & 0x000000ff) << 24) |
            ((bits & 0x0000ff00) << 8)  |
            ((bits & 0x00ff0000) >> 8)  |
            ((bits & 0xff000000) >> 24);
    return Uint32(shuffled);
}

void Encoder::Uint32(uint32_t v) {
    uint8_t buf[9];
    uint32_t space = 0x7f;
    uint8_t tag = 0;
    for (int o = 8; true; o--) {
        if (v <= space) {
            buf[o] = uint8_t(v) | uint8_t(tag);
            mOutput->write(buf + o, 9 - o);
            return;
        }
        buf[o] = uint8_t(v);
        v = v >> 8;
        space >>= 1;
        tag = (tag >> 1) | 0x80;
    }
}

void Encoder::Int32(int32_t v) {
    uint32_t uv = uint32_t(v) << 1;
    Uint32((v < 0) ? ~uv : uv);
}

void Encoder::Float64(double v) {
    uint64_t bits = *reinterpret_cast<uint64_t*>(&v);
    uint64_t shuffled =
            ((bits & 0x00000000000000ffULL) << 56) |
            ((bits & 0x000000000000ff00ULL) << 40) |
            ((bits & 0x0000000000ff0000ULL) << 24) |
            ((bits & 0x00000000ff000000ULL) << 8)  |
            ((bits & 0x000000ff00000000ULL) >> 8)  |
            ((bits & 0x0000ff0000000000ULL) >> 24) |
            ((bits & 0x00ff000000000000ULL) >> 40) |
            ((bits & 0xff00000000000000ULL) >> 56);
    return Uint64(shuffled);
}

void Encoder::Uint64(uint64_t v) {
    uint8_t buf[9];
    uint64_t space = 0x7f;
    uint8_t tag = 0;
    for (int o = 8; true; o--) {
        if (v <= space) {
            buf[o] = uint8_t(v) | uint8_t(tag);
            mOutput->write(buf + o, 9 - o);
            return;
        }
        buf[o] = uint8_t(v);
        v = v >> 8;
        space >>= 1;
        tag = (tag >> 1) | 0x80;
    }
}

void Encoder::Int64(int64_t v) {
    uint64_t uv = uint64_t(v) << 1;
    Uint64((v < 0) ? ~uv : uv);
}

void Encoder::Pointer(const void* p) {
    Uint64(reinterpret_cast<uintptr_t>(p));
    Uint32(0); // Pool-id
}

void Encoder::String(const char* v) {
    uint32_t len = v != nullptr ? static_cast<uint32_t>(strlen(v)) : 0;
    Uint32(len);
    mOutput->write(v, len);
}

void Encoder::Data(const void* ptr, int32_t size) {
    mOutput->write(ptr, size);
}

void Encoder::Entity(const schema::Entity* entity) {
    auto ret = mEntities.insert(
        std::make_pair(entity, mEntities.size()));
    uint32_t sid = ret.first->second;
    if (!ret.second) {
        Uint32(sid << 1);
    } else {
        Uint32((sid << 1) | 1);
        entity->encode(*this);
    }
}

void Encoder::Struct(const Encodable& obj) {
    obj.Encode(this);
}

void Encoder::Variant(const Encodable* obj) {
    if (obj == nullptr) {
        Entity(nullptr);
        return;
    }
    Entity(obj->Schema());
    obj->Encode(this);
}

void Encoder::Object(const Encodable* obj) {
    if (obj == nullptr) {
        Uint32(0);
        return;
    }
    uint32_t sid = ++mLastObjectId;
    Uint32((sid << 1) | 1);
    Variant(obj);
}

}  // namespace core

File: gapic/src/rpclib/com/google/gapid/rpclib/binary/Encoder.java
/*
 * Copyright (C) 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.google.gapid.rpclib.binary;

import com.google.common.collect.Maps;
import com.google.gapid.rpclib.schema.Entity;

import java.io.IOException;
import java.io.OutputStream;
import java.io.UnsupportedEncodingException;
import java.util.Map;

/**
 * An encoder of various primitive types.
 * The encoding format is documented at the following link:
 * https://github.com/google/gapid/+/master/binary/doc.go
 */
public class Encoder {
  private final OutputStream mOutputStream;
  private final Map<Entity, Integer> mEntities;
  private final Map<BinaryObject, Integer> mObjects;
  private final byte[] mBuffer;
  private boolean mControlNeeded = false;
  private final EncodingControl mControl = new EncodingControl();

  public Encoder(OutputStream out) {
    mEntities = Maps.newHashMap();
    mObjects = Maps.newHashMap();
    mOutputStream = out;
    mBuffer = new byte[9];
  }

  public void write(byte[] b, int len) throws IOException {
    mOutputStream.write(b, 0, len);
  }

  public void bool(boolean v) throws IOException {
    mBuffer[0] = (byte)(v ? 1 : 0);
    mOutputStream.write(mBuffer, 0, 1);
  }

  public void int8(byte v) throws IOException {
    mBuffer[0] = v;
    mOutputStream.write(mBuffer, 0, 1);
  }

  public void uint8(short v) throws IOException {
    mBuffer[0] = (byte)(v & 0xff);
    mOutputStream.write(mBuffer, 0, 1);
  }

  private void intv(long v) throws IOException {
    long uv = v << 1;
    if (v < 0) uv = ~uv;
    uintv(uv);
  }

  private void uintv(long v) throws IOException {
    long space = ~0x7fL;
    int tag = 0;
    for (int o = 8; true; o--) {
      if ((v & space) == 0) {
        mBuffer[o] = (byte)(v | tag);
        mOutputStream.write(mBuffer, o, 9 - o);
        return;
      }
      mBuffer[o] = (byte)(v&0xff);
      v >>>= 8;
      space >>= 1;
      tag =(tag >> 1) | 0x80;
    }
  }

  public void int16(short v) throws IOException {
    intv(v);
  }

  public void uint16(int v) throws IOException {
    uintv(v);
  }

  public void int32(int v) throws IOException {
    intv(v);
  }

  public void uint32(long v) throws IOException {
    uintv(v);
  }

  public void int64(long v) throws IOException {
    intv(v);
  }

  public void uint64(long v) throws IOException {
    uintv(v);
  }

  public void float32(float v) throws IOException {
    int bits = Float.floatToIntBits(v);
    int shuffled = ((bits & 0x000000ff) <<  24) |
                   ((bits & 0x0000ff00) <<   8) |
                   ((bits & 0x00ff0000) >> 8) |
                   ((bits & 0xff000000) >>> 24);
    uintv(shuffled);
  }

  public void float64(double v) throws IOException {
    long bits = Double.doubleToLongBits(v);
    long shuffled = ((bits & 0x00000000000000ffL) <<  56) |
                    ((bits & 0x000000000000ff00L) <<  40) |
                    ((bits & 0x0000000000ff0000L) <<  24) |
                    ((bits & 0x00000000ff000000L) <<   8) |
                    ((bits & 0x000000ff00000000L) >>   8) |
                    ((bits & 0x0000ff0000000000L) >>  24) |
                    ((bits & 0x00ff000000000000L) >> 40) |
                    ((bits & 0xff00000000000000L) >>> 56);
    uintv(shuffled);
  }

  public void string(String v) throws IOException {
    try {
      if (v == null) {
        uint32(0);
        return;
      }

      byte[] bytes = v.getBytes("UTF-8");
      uint32(bytes.length);
      for (byte b : bytes) {
        int8(b);
      }
    }
    catch (UnsupportedEncodingException e) {
      throw new RuntimeException(e); // Should never happen
    }
  }

  public void nonCompactString(String v) throws IOException {
    if (mControl.mode != EncodingControl.Compact) {
      string(v);
    }
  }

  public void writeSid(int sid, boolean encoded) throws IOException {
    if (mControlNeeded) {
      mControlNeeded = false; // set early to prevent recursive triggering
      uint32(1);             // encoded sid 0 is a special marker
      mControl.encode(this);      // and write the control block itself
    }
    uint32((sid << 1) | (encoded ? 1 : 0));
  }

  public void entity(Entity entity) throws IOException {
    if (entity == null) {
      writeSid(0, false);
    } else if (mEntities.containsKey(entity)) {
      writeSid(mEntities.get(entity), false);
    } else {
      int sid = mEntities.size() + 1;
      mEntities.put(entity, sid);
      writeSid(sid, true);
      entity.encode(this);
    }
  }

  public void value(BinaryObject obj) throws IOException {
    obj.klass().encode(this, obj);
  }

  public void variant(BinaryObject obj) throws IOException {
    if (obj == null) {
      entity(null);
    } else {
      BinaryClass c = obj.klass();
      entity(c.entity());
      c.encode(this, obj);
    }
  }

  public void object(BinaryObject obj) throws IOException {
    if (obj == null) {
      writeSid(0, false);
    } else if (mObjects.containsKey(obj)) {
      writeSid(mObjects.get(obj), false);
    } else {
      int sid = mObjects.size() + 1;
      mObjects.put(obj, sid);
      writeSid(sid, true);
      variant(obj);
    }
  }

  public OutputStream stream() {
    return mOutputStream;
  }

  public int getMode() {
    return mControl.mode;
  }

  public int setMode(int mode) {
    int oldMode = mControl.mode;
    if (oldMode != mode) {
      mControlNeeded = true;
      mControl.mode = mode;
    }
    return oldMode;
  }
}
